File: backend/models/backtest.py
"""
Backtest-related Pydantic models (Pydantic v2 compatible)
Consolidated from multiple model files for consistency
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class BacktestRequest(BaseModel):
    """
    バックテスト実行用のリクエストモデル。

    フロントエンドからの POST /simulate リクエストで使用される。
    """

    symbol: str = Field(..., description="Symbol code, e.g. AAPL, BTC/USDT")
    timeframe: str = Field(default="1d", description="Timeframe, e.g. 1d, 1h, 5m")

    start_date: Optional[str] = Field(
        default=None,
        description="Start datetime for backtest (ISO8601 string)",
    )
    end_date: Optional[str] = Field(
        default=None,
        description="End datetime for backtest (ISO8601 string)",
    )

    strategy: str = Field(
        default="ma_cross",
        description="Strategy name (currently only 'ma_cross' supported)",
    )

    initial_capital: float = Field(
        default=1_000_000,
        description="Initial capital for the backtest",
    )
    commission_rate: float = Field(
        default=0.0,
        description="Commission rate per trade (e.g., 0.001 = 0.1%)",
    )
    position_size: float = Field(
        default=1.0,
        description="Position size multiplier (0.0-1.0)",
    )

    short_window: int = Field(
        default=9,
        description="Short moving average window for MA cross strategy",
    )
    long_window: int = Field(
        default=21,
        description="Long moving average window for MA cross strategy",
    )


class EquityCurvePoint(BaseModel):
    """
    エクイティカーブ（残高推移）の1点。
    """

    date: str = Field(..., description="Time in ISO8601 format")
    equity: float = Field(..., description="Account equity at this timestamp")
    cash: Optional[float] = Field(default=None, description="Cash balance (optional)")


class TradeSummary(BaseModel):
    """
    バックテスト結果としての 1 トレード情報。
    """

    date: str = Field(..., description="Trade execution time (ISO8601)")
    side: str = Field(..., description="BUY or SELL")
    price: float = Field(..., description="Execution price")
    quantity: float = Field(..., description="Quantity traded")
    commission: float = Field(..., description="Commission paid")
    pnl: Optional[float] = Field(default=None, description="Profit/Loss (null for BUY)")
    cash_after: float = Field(..., description="Cash balance after trade")


class BacktestStats(BaseModel):
    """
    バックテスト全体の集計メトリクス。
    """

    initial_capital: float = Field(..., description="Initial capital")
    final_equity: float = Field(..., description="Final equity")
    total_pnl: float = Field(..., description="Total profit/loss")
    return_pct: float = Field(..., description="Return percentage")

    trade_count: int = Field(default=0, description="Total number of trades")
    winning_trades: int = Field(default=0, description="Number of winning trades")
    losing_trades: int = Field(default=0, description="Number of losing trades")
    win_rate: float = Field(default=0.0, description="Win rate (0.0-1.0)")

    max_drawdown: float = Field(default=0.0, description="Maximum drawdown")
    sharpe_ratio: Optional[float] = Field(default=None, description="Sharpe ratio")


class BacktestResponse(BaseModel):
    """
    /simulate バックテスト API のレスポンスモデル。
    """

    symbol: str = Field(..., description="Symbol backtested")
    timeframe: str = Field(..., description="Timeframe used")
    strategy: str = Field(..., description="Strategy name")

    equity_curve: List[EquityCurvePoint] = Field(
        default_factory=list,
        description="Equity curve data",
    )
    trades: List[TradeSummary] = Field(
        default_factory=list,
        description="List of all trades",
    )
    metrics: BacktestStats = Field(..., description="Summary statistics")

    data_points: int = Field(
        default=0,
        description="Number of candles used in backtest",
    )


__all__ = [
    "BacktestRequest",
    "BacktestResponse",
    "BacktestStats",
    "EquityCurvePoint",
    "TradeSummary",
]

File: frontend/src/api/backtest.ts
// frontend/src/api/backtest.ts
/// <reference types="vite/client" />

// ======================
// 型定義
// ======================

export interface BacktestRequest {
    symbol: string;
    timeframe?: string;        // e.g. "1d", "1h"
    start_date?: string | undefined;       // ISO 形式 "2020-01-01"
    end_date?: string | undefined;         // ISO 形式
    strategy?: string;         // "MA_CROSS" など（今は未使用でもOK）

    initial_capital?: number;  // 初期資金
    commission?: number;       // 手数料率
    position_size?: number;    // 1.0 = フルポジ

    short_window?: number;     // MA短期
    long_window?: number;      // MA長期
}

export interface BacktestMetrics {
    initial_capital: number;
    final_equity: number;
    total_pnl: number;
    return_pct: number;
    max_drawdown: number;
    win_rate: number;
    trade_count: number;
    winning_trades: number;
    losing_trades: number;
    sharpe_ratio?: number;
}

export interface BacktestTrade {
    date: string;              // ISO datetime string
    side: 'BUY' | 'SELL';
    price: number;
    quantity: number;
    commission: number;
    pnl?: number | null;       // BUY時はnull
    cash_after?: number;       // Python側であれば付けているはず（なければ undefined でOK）
}

export interface EquityPoint {
    date: string;              // ISO datetime string
    equity: number;
    cash?: number;
}

export interface BacktestResponse {
    metrics: BacktestMetrics;
    trades: BacktestTrade[];
    equity_curve: EquityPoint[];
}

// ===== Experiments =====

export interface BacktestExperimentCreate {
    name: string;
    description?: string;
    request: BacktestRequest;
    result?: BacktestResponse | null;
}

export interface BacktestExperiment {
    id: string;
    name: string;
    description?: string;
    created_at: string;
    updated_at: string;
    request: BacktestRequest;
    result?: BacktestResponse | null;
}

export interface BacktestExperimentSummary {
    id: string;
    name: string;
    description?: string;
    created_at: string;
    updated_at: string;
}

// ======================
// API ベース URL
// ======================

const API_BASE = import.meta.env.VITE_API_BASE ?? 'http://localhost:8000';

// ======================
// バックテスト実行
// ======================

export async function runBacktest(
    request: BacktestRequest,
): Promise<BacktestResponse> {
    const response = await fetch(`${API_BASE}/simulate`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
    });

    if (!response.ok) {
        let detail = response.statusText;
        try {
            const err = await response.json();
            if (err?.detail) detail = err.detail;
        } catch {
            // ignore
        }
        throw new Error(detail || 'Simulation failed');
    }

    return response.json();
}

// ======================
// Experiment API 群
// ======================

export async function createExperiment(
    payload: BacktestExperimentCreate,
): Promise<BacktestExperiment> {
    const res = await fetch(`${API_BASE}/experiments`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
    });

    if (!res.ok) {
        let detail = res.statusText;
        try {
            const err = await res.json();
            if (err?.detail) detail = err.detail;
        } catch {
            // ignore
        }
        throw new Error(detail || 'Create experiment failed');
    }

    return res.json();
}

export async function listExperiments(): Promise<BacktestExperimentSummary[]> {
    const res = await fetch(`${API_BASE}/experiments`);

    if (!res.ok) {
        let detail = res.statusText;
        try {
            const err = await res.json();
            if (err?.detail) detail = err.detail;
        } catch {
            // ignore
        }
        throw new Error(detail || 'List experiments failed');
    }

    return res.json();
}

export async function getExperiment(
    id: string,
): Promise<BacktestExperiment> {
    const res = await fetch(`${API_BASE}/experiments/${id}`);

    if (!res.ok) {
        let detail = res.statusText;
        try {
            const err = await res.json();
            if (err?.detail) detail = err.detail;
        } catch {
            // ignore
        }
        throw new Error(detail || 'Get experiment failed');
    }

    return res.json();
}

export async function runExperiment(
    id: string,
): Promise<BacktestResponse> {
    const res = await fetch(`${API_BASE}/experiments/${id}/run`, {
        method: 'POST',
    });

    if (!res.ok) {
        let detail = res.statusText;
        try {
            const err = await res.json();
            if (err?.detail) detail = err.detail;
        } catch {
            // ignore
        }
        throw new Error(detail || 'Run experiment failed');
    }

    return res.json();
}

// ======================
// ヘルパー（任意）
// ======================

export function createDefaultBacktestRequest(symbol: string): BacktestRequest {
    return {
        symbol,
        timeframe: '1d',
        initial_capital: 1_000_000,
        commission: 0.001,
        position_size: 1.0,
        short_window: 9,
        long_window: 21,
    };
}

// 既存の export function runExperiment(...) { ... } のあととかでOK

// 互換用ラッパー：既存コードが runSimulation を呼んでいても動くようにする
export async function runSimulation(
    request: BacktestRequest,
): Promise<BacktestResponse> {
    return runBacktest(request);
}
